packages/components/Shimmer/src/Shimmer.tsx

```diff
@@ -0,0 +1,138 @@
+/** @jsx withSlots */
+import { compose, UseSlots, buildProps, mergeProps, withSlots } from '@fluentui-react-native/framework';
+import { processColor, ViewProps } from 'react-native';
+import { IViewProps } from '@fluentui-react-native/adapters';
+import * as React from 'react';
+import { ensureNativeComponent } from '@fluentui-react-native/component-cache';
+
+const shimmerName = 'Shimmer';
+
+const NativeShimmerView = ensureNativeComponent('MSFShimmerView');
+
+/**
+ * Appearance of the views that are shimmered
+ */
+export interface ShimmerViewAppearance {
+  /**
+   * Color to tint the shimmer boxes. Defaults to the Fabric default color.
+   */
+  tintColor?: number;
+
+  /**
+   * Corner radius on each view.
+   */
+  cornerRadius?: number;
+
+  /**
+   * Corner radius on each UILabel. Set to < 0 to disable and use default `cornerRadius`.
+   */
+  labelCornerRadius?: number;
+
+  /**
+   * usesTextHeightForLabels: True to enable shimmers to auto-adjust to font height for a UILabel -- this
+   * will more accurately reflect the text in the label rect rather than using the bounding box. `labelHeight`
+   * will take precedence over this property.
+   */
+  usesTextHeightForLabels?: boolean;
+
+  /**
+   * If greater than 0, a fixed height to use for all UILabels. This will take precedence
+   * over `usesTextHeightForLabels`. Set to less than 0 to disable.
+   */
+  labelHeight?: number;
+}
+
+/**
+ * Appearance of the shimmer itself (the animation appearance)
+ */
+export interface ShimmerAppearance {
+  alpha?: number;
+  width?: number;
+
+  /**
+   * Angle of the direction of the gradient, in radian. 0 means horizontal, Pi/2 means vertical.
+   */
+  angle?: number;
+
+  /**
+   * Speed of the animation, in point/seconds.
+   */
+  speed?: number;
+
+  /**
+   * Delay between the end of a shimmering animation and the beginning of the next one, in seconds
+   */
+  delay?: number;
+}
+
+/**
+ * ViewProps props, based off of the standard react-native ViewProps with some new extensions
+ */
+export type NativeShimmerViewProps<TBase = IViewProps> = TBase & {
+  /**
+   * Appearance of the views that are shimmered. If not passed, the default will be used.
+   */
+  appearance?: ShimmerViewAppearance;
```

Comment:
A few questions here for you:

- do we want these to be settable via props or do we want people to use customization to set them?
- should they be flat to match tokens?

Some points to understand what is happening as you have it written.

- If they set tokens via customize, or just use normally, the props passed to the native component will have two objects, appearance and shimmerAppearance. These will be created once and will have consistent object references.
- If react does a shallow compare to see if props have changed it will see they are the same object and not think it needs to do any remounts.
- If someone sets these as props, these objects will be not be merged and will fully replace what is in tokens. This is because mergeProps only does a deep merge on the style property.
- These will have the same pitfalls as setting styles, inline declarations will cause new objects to be created. Style has lint rules typically associated with it to catch them but these will be up to users to use correctly.

It would work better to do one of the following:

- remove these from props and just tell people to use customize to change them. This ensures they get good caching and shouldn't require any other adjustments.
- make the props flat, similar to what was done in tokens, then set tokenProps to true in the framework, this will patch your token values with any values set in props and the tokens will be present by the time your buildProps routine runs. You will then need to use the array parameter for buildProps to declare what values it should cache on.
- alternatively you can expose a subset in props that you expect people to modify more commonly. In this case you would follow the steps above except tokenProps would be set to just have that list of props that is both a prop and a token.

Reply:
One other point with the current behavior, if someone only wants to change one property, like corner radius, they still have to set all of the values. This is harder to use and requires them to look up the right constants to copy. The downside here is that if we change those their code won't update to match.

Reply:
As of right now, I expect all of the members of appearance and shimmerAppearance to only be set once at creation of the shimmer (making it more suited as tokens than props). I have those two as not flat to match the API on the native API: the iOS (and eventually android and macOS) controlc each have two structs called appearance and shimmerAppearance that 1:1 map with what I exposed in JS.

I thought that by putting appearance & shimmerAppearance in NativeShimmerViewProps, clients would not be able to manually set those as props, and would only be able to use the token. By your 1st suggestion, do you mean to remove the NativeShimmerViewProps from line 132?

The 2nd suggestion is probably the change we want in the native API long term anyway. The non flat list of properties is weird IMO.

For having to set all the properties, would that be fixed by adding "?"'s to each token in lines 98-107?

Reply:
I made changes to the native side that should address your comments by taking your 2nd suggestion. The Native ShimmerView now has a flat list of properties it takes in, and the React-Native tokenprops match 1:1 with them. We don't need to explicitly set all of them to stylize the shimmerview, and the default / stylized shimmer views work as expected.